File: webflux-dialogue/src/main/java/com/study/webflux/rag/domain/dialogue/model/ConversationTurn.java
package com.study.webflux.rag.domain.dialogue.model;

import java.time.Instant;

public record ConversationTurn(
	String id,
	ConversationSessionId sessionId,
	String query,
	String response,
	Instant createdAt
) {
	public ConversationTurn {
		if (sessionId == null) {
			throw new IllegalArgumentException("sessionId cannot be null");
		}
		if (query == null || query.isBlank()) {
			throw new IllegalArgumentException("query cannot be null or blank");
		}
		if (createdAt == null) {
			createdAt = Instant.now();
		}
	}

	public static ConversationTurn create(ConversationSessionId sessionId, String query) {
		return new ConversationTurn(null, sessionId, query, null, Instant.now());
	}

	public static ConversationTurn withId(String id,
		ConversationSessionId sessionId,
		String query,
		String response,
		Instant createdAt) {
		return new ConversationTurn(id, sessionId, query, response, createdAt);
	}

	public ConversationTurn withResponse(String response) {
		return new ConversationTurn(this.id, this.sessionId, this.query, response, this.createdAt);
	}
}

File: domain/src/main/kotlin/com/miyou/app/domain/dialogue/model/ConversationTurn.kt
package com.miyou.app.domain.dialogue.model

import java.time.Instant

data class ConversationTurn(
    val id: String?,
    val sessionId: ConversationSessionId,
    val query: String,
    val response: String?,
    val createdAt: Instant,
) {
    init {
        require(query.isNotBlank()) { "query cannot be null or blank" }
    }

    fun withResponse(response: String?): ConversationTurn = copy(response = response)

    fun id(): String? = id

    fun sessionId(): ConversationSessionId = sessionId

    fun query(): String = query

    fun response(): String? = response

    fun createdAt(): java.time.Instant = createdAt

    companion object {
        fun create(
            sessionId: ConversationSessionId,
            query: String,
        ): ConversationTurn = ConversationTurn(null, sessionId, query, null, Instant.now())

        fun withId(
            id: String?,
            sessionId: ConversationSessionId,
            query: String,
            response: String?,
            createdAt: Instant,
        ): ConversationTurn = ConversationTurn(id, sessionId, query, response, createdAt)
    }
}
